webapp/channels/src/components/convert_gm_to_channel_modal/convert_gm_to_channel_modal.tsx

```diff
@@ -0,0 +1,220 @@
+import {trackEvent} from 'actions/telemetry_actions';
+
+import ChannelNameFormField from 'components/channel_name_form_field/channel_name_form_field';
+import type {Actions} from 'components/convert_gm_to_channel_modal/index';
+import NoCommonTeamsError from 'components/convert_gm_to_channel_modal/no_common_teams/no_common_teams';
+import TeamSelector from 'components/convert_gm_to_channel_modal/team_selector/team_selector';
+import WarningTextSection from 'components/convert_gm_to_channel_modal/warning_text_section/warning_text_section';
+import LoadingSpinner from 'components/widgets/loading/loading_spinner';
+
+export type Props = {
+    onExited: () => void;
+    channel: Channel;
+    actions: Actions;
+    profilesInChannel: UserProfile[];
+    teammateNameDisplaySetting: string;
+    channelsCategoryId: string | undefined;
+    currentUserId: string;
+}
+
+const ConvertGmToChannelModal = (props: Props) => {
+    const intl = useIntl();
+    const {formatMessage} = intl;
+
+    const [channelName, setChannelName] = useState<string>('');
+    const handleChannelNameChange = useCallback((newName: string) => {
+        setChannelName(newName);
+    }, []);
+
+    const [channelURL, setChannelChannelURL] = useState<string>('');
+    const handleChannelURLChange = useCallback((newURL: string) => {
+        setChannelChannelURL(newURL);
+    }, []);
+
+    const [channelMemberNames, setChannelMemberNames] = useState<string[]>([]);
+
+    useEffect(() => {
+        const validProfilesInChannel = props.profilesInChannel.
+            filter((user) => user.id !== props.currentUserId && user.delete_at === 0).
+            map((user) => displayUsername(user, props.teammateNameDisplaySetting));
+
+        setChannelMemberNames(validProfilesInChannel);
+    }, [props.profilesInChannel]);
+
+    const [commonTeamsById, setCommonTeamsById] = useState<{[id: string]: Team}>({});
+    const [commonTeamsFetched, setCommonTeamsFetched] = useState<boolean>(false);
+    const [loadingAnimationTimeout, setLoadingAnimationTimeout] = useState<boolean>(false);
+    const [selectedTeamId, setSelectedTeamId] = useState<string>();
+    const [nameError, setNameError] = useState<boolean>(false);
+    const [conversionError, setConversionError] = useState<string>();
+
+    const handleTeamChange = useCallback((teamId: string) => {
+        setSelectedTeamId(teamId);
+    }, []);
+
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        const work = async () => {
+            const response = await dispatch(getGroupMessageMembersCommonTeams(props.channel.id)) as ActionResult<Team[], ServerError>;
+            if (response.error || !response.data) {
+                return;
+            }
+            const teams = response.data;
+
+            const teamsById: {[id: string]: Team} = {};
+            teams.forEach((team) => {
+                teamsById[team.id] = team;
+            });
```

Comment:
0/5 I like for this using a reduce, but probably the same.

Reply:
Technically, I think this'd result in fewer new objects being created than if you were to use reduce, but I doubt it's ever enough to matter in practice. Then again, I'm also a fan of `for (const team in teams)` myself 😛

Reply:
All paths are equally-ish long, and lead to same destination :P